### PR TITLE
[COST-4372] OCP: Fix capacity instance count

### PR DESCRIPTION
**Pull request by djnakabaale**
## Jira Ticket

[COST-4372](https://issues.redhat.com/browse/COST-4372)

## Description

This change fixes aggregation count for cluster capacity to count each node per cluster once.

## Testing

1. Checkout this branch
2. Restart Koku
3. Load test data with `make create-test-customer` and `make load-test-customer-data`
4. Hit these endpoints (should also expect similar behavior on `memory` and `volume` reports)

    - http://localhost:8000/api/cost-management/v1/reports/openshift/compute/?group_by[node]=*&filter[resolution]=monthly&filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[cluster]=Test OCP on Prem

        <details>
        <summary>
        You should see the total capacity count value in the meta equal to the sum of all node counts, similar to this:
        </summary>
            
            {
                "meta": {
                    "count": 1,
                    "limit": 100,
                    "offset": 0,
                    "currency": "USD",
                    "filter": {
                        "resolution": "monthly",
                        "time_scope_value": "-1",
                        "time_scope_units": "month",
                        "cluster": [
                            "Test OCP on Prem"
                        ]
                    },
                    "group_by": {
                        "node": [
                            "*"
                        ]
                    },

                    "total": {
                        ...
                        "capacity": {
                            "value": 13056.0,
                            "units": "Core-Hours",
                            "unused": 6936.0,
                            "unused_percent": 53.125,
                            "count": 32.0,
                            "count_units": "Core"
                        }
                        ...
                    }
                },
                ...
                "data": [
                    {
                        "date": "2024-03",
                        "nodes": [
                            {
                                "node": "compute_4",
                                "values": [
                                    {
                                        "node": "compute_4",
                                        "date": "2024-03",
                                        "usage": {
                                            "value": 430.71782,
                                            "units": "Core-Hours"
                                        },
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 816.0,
                                            "unused_percent": 50.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            },
                            {
                                "node": "compute_2",
                                "values": [
                                    {
                                        "node": "compute_2",
                                        "date": "2024-03",
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 816.0,
                                            "unused_percent": 50.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            },
                            {
                                "node": "compute_1",
                                "values": [
                                    {
                                        "node": "compute_1",
                                        "date": "2024-03",
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 816.0,
                                            "unused_percent": 50.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            },
                            {
                                "node": "compute_6",
                                "values": [
                                    {
                                        "node": "compute_6",
                                        "date": "2024-03",
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 816.0,
                                            "unused_percent": 50.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            },
                            {
                                "node": "compute_5",
                                "values": [
                                    {
                                        "node": "compute_5",
                                        "date": "2024-03",
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 816.0,
                                            "unused_percent": 50.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            },
                            {
                                "node": "master_3",
                                "values": [
                                    {
                                        "node": "master_3",
                                        "date": "2024-03",
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 816.0,
                                            "unused_percent": 50.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            },
                            {
                                "node": "compute_3",
                                "values": [
                                    {
                                        "node": "compute_3",
                                        "date": "2024-03",
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 816.0,
                                            "unused_percent": 50.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            },
                            {
                                "node": "master_1",
                                "values": [
                                    {
                                        "node": "master_1",
                                        "date": "2024-03",
                                        ...
                                        "capacity": {
                                            "value": 1632.0,
                                            "units": "Core-Hours",
                                            "unused": 1224.0,
                                            "unused_percent": 75.0,
                                            "count": 4.0
                                        }
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }
            
        </details>

    - http://localhost:8000/api/cost-management/v1/reports/openshift/compute/?group_by[cluster]=*&filter[resolution]=monthly&filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[cluster]=Test OCP on Prem

        <details>
        <summary>
        You should see the total capacity count value in the meta equal to the sum of all cluster counts, similar to this:
        </summary>

            "meta": {

                "filter": {
                    "resolution": "monthly",
                    "time_scope_value": "-1",
                    "time_scope_units": "month",
                    "cluster": [
                        "Test OCP on Prem"
                    ]
                },
                "group_by": {
                    "cluster": [
                        "Test OCP on Prem"
                    ]
                },

                "total": {
                    
                    "capacity": {
                        "value": 13056.0,
                        "units": "Core-Hours",
                        "unused": 6936.0,
                        "unused_percent": 53.125,
                        "count": 32.0,
                        "count_units": "Core"
                    }
                }
            },
            ...
            "data": [
                {
                    "date": "2024-03",
                    "clusters": [
                        {
                            "cluster": "my-ocp-cluster-3",
                            "values": [
                                {
                                    "date": "2024-03",
                                    "cluster": "my-ocp-cluster-3",
                                
                                    "capacity": {
                                        "value": 13056.0,
                                        "units": "Core-Hours",
                                        "unused": 6936.0,
                                        "unused_percent": 53.125,
                                        "count": 32.0,
                                        "count_units": "Core"
                                    },
                                    "clusters": [
                                        "Test OCP on Premises"
                                    ],
                                    "source_uuid": [
                                        "e3d22b65-7830-464a-ba23-d80e2cd1d1f2"
                                    ]
                                    ...
                                }
                            ]
                        }
                    ]
                }
            ]
        </details>

    - http://localhost:8000/api/cost-management/v1/reports/openshift/compute/?group_by[cluster]=*&filter[cluster]=Test OCP on Prem

        <details>
        <summary>
        You should see the total capacity count value in the meta equal to the sum of all cluster counts for each date, similar to this:
        </summary>

            {
                "meta": {
                    
                    "filter": {
                        "cluster": [
                            "Test OCP on Prem"
                        ],
                        "time_scope_value": "-10",
                        "time_scope_units": "day",
                        "resolution": "daily"
                    },
                    "group_by": {
                        "cluster": [
                            "Test OCP on Prem"
                        ]
                    },

                    "total": {
                        
                        "capacity": {
                            "value": 6912.0,
                            "units": "Core-Hours",
                            "unused": 3672.0,
                            "unused_percent": 53.125,
                            "count": 32.0,
                            "count_units": "Core"
                        },
                        
                    }
                },
                ...
                "data": [
                    {
                        "date": "2024-03-09",
                        "clusters": [
                            {
                                "cluster": "my-ocp-cluster-3",
                                "values": [
                                    {
                                        "date": "2024-03-09",
                                        "cluster": "my-ocp-cluster-3",
                                        
                                        "capacity": {
                                            "value": 768.0,
                                            "units": "Core-Hours",
                                            "unused": 408.0,
                                            "unused_percent": 53.125,
                                            "count": 32.0,
                                            "count_units": "Core"
                                        }
                                    }
                                ]
                            }
                        ]
                    },
                    {
                        "date": "2024-03-10",
                        "clusters": [
                            {
                                "cluster": "my-ocp-cluster-3",
                                "values": [
                                    {
                                        "date": "2024-03-10",
                                        "cluster": "my-ocp-cluster-3",
                                        
                                        "capacity": {
                                            "value": 768.0,
                                            "units": "Core-Hours",
                                            "unused": 408.0,
                                            "unused_percent": 53.125,
                                            "count": 32.0,
                                            "count_units": "Core"
                                        }
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }

        </details>

    - http://localhost:8000/api/cost-management/v1/reports/openshift/compute/?filter[resolution]=monthly&filter[cluster]=Test OCP on AWS, Test OCP on Premises&group_by[cluster]=*

        <details>
        <summary>
            You should see the total capacity count value in the meta equal to the sum of all cluster counts for per month. For this example the total count value meta should be equal to the sum of all cluster counts for "date": "2024-03", similar to this:
        </summary>

            {
                "meta": {

                    "filter": {
                        "resolution": "monthly",
                        "cluster": [
                            "Test OCP on AWS",
                            "Test OCP on Premises"
                        ],
                        "time_scope_value": "-10",
                        "time_scope_units": "day"
                    },
                    "group_by": {
                        "cluster": [
                            "Test OCP on AWS",
                            "Test OCP on Premises"
                        ]
                    },

                    "total": {
                        
                        "capacity": {
                            "value": 10368.0,
                            "units": "Core-Hours",
                            "unused": 4320.0,
                            "unused_percent": 41.666666666666664,
                            "count": 48.0,
                            "count_units": "Core"
                        },
                        
                    }
                },
                ...
                "data": [
                    {
                        "date": "2024-03",
                        "clusters": [
                            {
                                "cluster": "my-ocp-cluster-3",
                                "values": [
                                    {
                                        "date": "2024-03",
                                        "cluster": "my-ocp-cluster-3",
                                        
                                        "capacity": {
                                            "value": 6912.0,
                                            "units": "Core-Hours",
                                            "unused": 3672.0,
                                            "unused_percent": 53.125,
                                            "count": 32.0,
                                            "count_units": "Core"
                                        }
                                    }
                                ]
                            },
                            {
                                "cluster": "my-ocp-cluster-1",
                                "values": [
                                    {
                                        "date": "2024-03",
                                        "cluster": "my-ocp-cluster-1",
                                        
                                        "capacity": {
                                            "value": 3456.0,
                                            "units": "Core-Hours",
                                            "unused": 648.0,
                                            "unused_percent": 18.75,
                                            "count": 16.0,
                                            "count_units": "Core"
                                        }
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }

        </details>

## Release Notes
- [x] proposed release note

```markdown
* [COST-4372](https://issues.redhat.com/browse/COST-4732) Fix monthly cluster capacity count aggregation
```

### CODE REVIEW

**Comment by djnakabaale** [comment]
ocp-smoke-test: https://ci.ext.devshift.net/job/koku-pipeline-pr-check-main/1514/

File: koku/api/report/ocp/capacity/cluster_capacity.py [hunk 1]
```diff
@@ -83,28 +85,117 @@ def capacity_annotations(self):
...
+            if cluster_date_key not in daily_max_node_capacities:
+                daily_max_node_capacities[cluster_date_key] = {}
+
+            daily_max_node_capacities[cluster_date_key][node] = max(
+                capacity_count, daily_max_node_capacities[cluster_date_key].get(node, 0)
+            )
+
+        for (cluster, date), nodes in daily_max_node_capacities.items():
+            if date not in self.count_by_date_cluster:
+                self.count_by_date_cluster[date] = {}
```
Comment (maskarb) [inline koku/api/report/ocp/capacity/cluster_capacity.py:172]: `count_by_date_cluster`. is a defaultdict, so this check here is not necessary:
```suggestion
```

File: koku/api/report/ocp/capacity/cluster_capacity.py [hunk 2]
```diff
@@ -83,28 +85,117 @@ def capacity_annotations(self):
...
+            # count each node per cluster once
+            # use max capacity count if a node has multiple values
+            if cluster not in unique_cluster_node_counts:
+                unique_cluster_node_counts[cluster] = {node: capacity_count}
+            if node not in unique_cluster_node_counts[cluster]:
+                unique_cluster_node_counts[cluster][node] = capacity_count
+            else:
+                unique_cluster_node_counts[cluster][node] = max(
+                    capacity_count, unique_cluster_node_counts[cluster].get(node, 0)
+                )
```
Comment (maskarb) [inline koku/api/report/ocp/capacity/cluster_capacity.py:127]: I believe you can also eliminate a bunch of branches here by using a defaultdict for the `unique_cluster_node_counts`:
```suggestion
        unique_cluster_node_counts = defaultdict(lambda: defaultdict(Decimal))
        for count in node_instance_counts:
            cluster, node, capacity_count = (count["cluster"], count["node"], count["capacity_count"])

            # count each node per cluster once
            # use max capacity count if a node has multiple values
            unique_cluster_node_counts[cluster][node] = max(
                capacity_count, unique_cluster_node_counts[cluster].get(node, 0)
            )
```

File: koku/api/report/ocp/capacity/cluster_capacity.py [hunk 3]
```diff
@@ -83,28 +85,117 @@ def capacity_annotations(self):
...
+            usage_start_resolution = self._resolution_usage_converter(usage_start)
+
+            # composite key to access daily_max_node_capacities dict
+            cluster_date_key = (cluster, usage_start_resolution)
+            if cluster_date_key not in daily_max_node_capacities:
+                daily_max_node_capacities[cluster_date_key] = {}
+
+            daily_max_node_capacities[cluster_date_key][node] = max(
+                capacity_count, daily_max_node_capacities[cluster_date_key].get(node, 0)
+            )
```
Comment (maskarb) [inline koku/api/report/ocp/capacity/cluster_capacity.py:168]: Same as above, some of this branching can be eliminated by using a defaultdict:
```suggestion
        daily_max_node_capacities = defaultdict(lambda: defaultdict(Decimal))
        for count in node_instance_counts:
            cluster, node, usage_start, capacity_count = (
                count["cluster"],
                count["node"],
                count["usage_start"],
                count["capacity_count"],
            )
            # count each node per cluster per date once
            # use max capacity count if a node has multiple values
            usage_start_resolution = self._resolution_usage_converter(usage_start)

            # composite key to access daily_max_node_capacities dict
            cluster_date_key = (cluster, usage_start_resolution)

            daily_max_node_capacities[cluster_date_key][node] = max(
                capacity_count, daily_max_node_capacities[cluster_date_key].get(node, 0)
            )
```

File: koku/api/report/ocp/capacity/node_capacity.py [hunk 4]
```diff
@@ -70,13 +66,23 @@ def populate_dataclass(self):
         Creates and populates the capacity dataclass.
         """
         cap_key = list(self.capacity_annotations.keys())[0]
-        cap_data = self.query.values(*["usage_start", "node"]).annotate(**self.capacity_annotations)
-        for entry in cap_data:
-            node_key = entry.get("node", "")
+        node_capacity_counts = self.query.values(*["usage_start", "node"]).annotate(**self.capacity_annotations)
+
+        # Count each node once. Use max capacity count if a node has multiple values
+        unique_nodes = {}
```
Comment (maskarb) [inline koku/api/report/ocp/capacity/node_capacity.py:72]:
```suggestion
        unique_nodes = defaultdict(Decimal)
```

File: koku/api/report/ocp/capacity/node_capacity.py [hunk 5]
```diff
@@ -70,13 +66,23 @@ def populate_dataclass(self):
...
+            capacity_count = entry.get("capacity_count", 0)
             usage_start = self._resolution_usage_converter(entry.get("usage_start", ""))
             cap_value = entry.get(cap_key, 0)
-            self._add_capacity(cap_value, usage_start, node_key)
-            self._add_count(entry.get("capacity_count", 0))
+            self._add_capacity(cap_value, usage_start, node)
+
+            if node not in unique_nodes:
+                unique_nodes[node] = capacity_count
+            unique_nodes[node] = max(capacity_count, unique_nodes.get(node, 0))
```
Comment (maskarb) [inline koku/api/report/ocp/capacity/node_capacity.py:82]: Same here if we use a defaultdict above:
```suggestion
            unique_nodes[node] = max(capacity_count, unique_nodes.get(node, 0))
```

File: koku/api/report/ocp/capacity/cluster_capacity.py [hunk 6]
```diff
@@ -9,15 +9,17 @@
...
 
 
 # Developer Note:
 # This is common code used in the
 # query handler and node capacity.
-def calculate_unused(row, finalized_mapping={}):
+def calculate_unused(row, finalized_mapping=None):
     """Calculates the unused portions of the capacity & request."""
     # Populate unused request and capacity
+    finalized_mapping = finalized_mapping if finalized_mapping else {}
```
Comment (maskarb) [inline koku/api/report/ocp/capacity/cluster_capacity.py:22]: This eliminates branching:
```suggestion
    finalized_mapping = finalized_mapping or {}
```

```
>>> None or {}
{}
```
```

**Review by maskarb — COMMENTED** [review]
This looks good! Here are some more suggestions:

**Review by maskarb — APPROVED** [review]
lgtm!

**Pull request merged**